docs: add getting started tutorial for dotnet (#1174)
ory/docs: docs/getting-started/integrate-auth/dotnet.mdx

---
id: dotnet
title: Integrate authentication into ASP.NET Core
sidebar_label: Dotnet / ASP.NET Core
---

```mdx-code-block
import CodeBlock from '@theme/CodeBlock'
import Teaser from '../_common/teaser.mdx'
import mp4 from '../_static/go/screencast.mp4'
import webm from '../_static/go/screencast.webm'

<Teaser
  language={<a href="https://dotnet.microsoft.com/en-us/">.NET</a>}
  framework={<a href="https://dotnet.microsoft.com/en-us/apps/aspnet">ASP.NET Core</a>}
  mp4={mp4}
  webm={webm}
  url="https://github.com/ory/docs/tree/master/code-examples/protect-page-login/dotnet/01-basic"
/>
```

## Install Ory CLI

Follow [this guide](../../guides/cli/01_installation.mdx) to install the Ory CLI on your machine.

### Why do I Need the Ory CLI?

```mdx-code-block
import OryCLI from '../_common/ory-cli.mdx'

<OryCLI />
```

### Usage of Ory Tunnel

> For this example we will be using the [Ory Tunnel](../../guides/cli/proxy-and-tunnel#ory-tunnel). The interactions beeing

Reviewer: > a
```suggestion
We will be using the [Ory Tunnel](../../guides/cli/proxy-and-tunnel#ory-tunnel) in this tutorial.
```

### Create an ASP.NET Core app

We will base this on the empty web template. Run these commands:

```shell-session
mkdir your-project
cd your-project

dotnet new web
```

![Create ASP.NET Core app](../_static/dotnet/new_project.gif)

### Install Ory SDK

Run this command to install the Ory SDK which allows you to interact with Ory APIs:

```shell-session
dotnet add package Ory.Client
```

![Add Ory.Client package](../_static/dotnet/add_package.gif)

### Configure the SDK

We will need to add the Ory Client to this app. Make the following changes to the `Program.cs` file that was generated:

```csharp title="Program.cs" showLineNumbers {4-9}
var builder = WebApplication.CreateBuilder(args);
var app = builder.Build();

// create a new Ory Client with the BasePath set to the Ory Tunnel enpoint
var oryBasePath = builder.Configuration.GetValue<string>("ORY_BASEPATH") ?? "http://localhost:4000";
var ory = new FrontendApi(new Configuration
{
	BasePath = oryBasePath
});

app.MapGet("/", () => "Hello World!");

app.Run();
```

## Validate and login

Next, create middleware that checks if the user has a valid session. The session is considered valid when the user is
authenticated.

> The middleware passes the current `HttpContext` cookies to the Ory client to check if the session is valid.

Reviewer: ```suggestion
The middleware passes the request cookies to the Ory client to check if the session is valid.
If the session is valid and active, the session information is recorded in `ctx`.
```

}

<html lang="en">
  <head>
    <title>Ory Network secured ASP.NET Core web app</title>
  </head>
  <body>
    <h1>Dashboard</h1>
    <hr />
    <h2>Your Session Data:</h2>
    <pre><code>@session</code></pre>
  </body>
</html>
```

### App configuration

To configure the app to serve RazorPages, make the following changes in `Program.cs`:

```csharp title="Program.cs" showLineNumbers {3-4,21-22}
var builder = WebApplication.CreateBuilder(args);

// add support for RazorPages
builder.Services.AddRazorPages();

var app = builder.Build();

// create a new Ory Client with the BasePath set to the Ory Tunnel enpoint
var oryBasePath = builder.Configuration.GetValue<string>("ORY_BASEPATH") ?? "http://localhost:4000";
var ory = new FrontendApi(new Configuration
{
	BasePath = oryBasePath
});

// add session middleware
app.Use(async (ctx, next) =>
{
	// ... session middleware logic ...
});

// configure pipeline to use RazorPages
app.MapRazorPages();

app.Run();
```

## Test your application

With all of the pieces in place, it's time to test your application. Follow these steps:

> 1. Start your ASP.NET app:

Reviewer: I followed the tutorial exactly to this point and get this:

```
$ dotnet run
Building...
/Users/arneluenser/dev/dotnet-tutorial/Program.cs(9,15): error CS0246: The type or namespace name 'FrontendApi' could not be found (are you missing a using directive or an assembly reference?) [/Users/arneluenser/dev/dotnet-tutorial/dotnet-tutorial.csproj]
/Users/arneluenser/dev/dotnet-tutorial/Program.cs(9,31): error CS0246: The type or namespace name 'Configuration' could not be found (are you missing a using directive or an assembly reference?) [/Users/arneluenser/dev/dotnet-tutorial/dotnet-tutorial.csproj]
/Users/arneluenser/dev/dotnet-tutorial/Program.cs(35,9): error CS0246: The type or namespace name 'ApiException' could not be found (are you missing a using directive or an assembly reference?) [/Users/arneluenser/dev/dotnet-tutorial/dotnet-tutorial.csproj]

The build failed. Fix the build errors and run again.
```

Author: I have left out the namespace imports in the code boxes on purpose to remove noise. Maybe they should be added to make the tutorial truly self contained?

It's only two lines of Code, so I will add the imports.
